ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -0,0 +1,205 @@
+type RayJobReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+}
+
+// NewRayJobReconciler returns a new reconcile.Reconciler
+func NewRayJobReconciler(mgr manager.Manager) *RayJobReconciler {
+	return &RayJobReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+		Log:    ctrl.Log.WithName("controllers").WithName("RayJob"),
+	}
+}
+
+// Reconcile reads that state of a RayJob object and makes changes based on it
+// and what is in the RayJob.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write workloads
+// +kubebuilder:rbac:groups=ray.io,resources=rayjobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ray.io,resources=rayjobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=ray.io,resources=rayjobs/finalizer,verbs=update
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=pods/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=coordination.k8s.io,resources=leases,verbs=get;list;create;update
+// +kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=roles,verbs=get;list;watch;create;delete;update
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=get;list;watch;create;delete
+// Reconcile used to bridge the desired state with the current state
+func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("rayjob", request.NamespacedName)
+	log.Info("reconciling RayJob", "service NamespacedName", request.NamespacedName)
+
+	// Get serving cluster instance
+	var err error
+	var RayJobInstance *rayv1alpha1.RayJob
+
+	if RayJobInstance, err = r.getRayJobInstance(ctx, request); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var rayClusterInstance *rayv1alpha1.RayCluster
+	if rayClusterInstance, err = r.getOrCreateRayClusterInstance(ctx, RayJobInstance); err != nil {
+		err = r.updateState(ctx, RayJobInstance, rayv1alpha1.FailToGetOrCreateRayCluster, err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log.Info("Updated RayCluster")
+
+	RayJobInstance.Status.RayClusterStatus = rayClusterInstance.Status
+
+	var clientURL string
+	if clientURL, err = r.fetchDashboardURL(ctx, rayClusterInstance); err != nil || clientURL == "" {
+		err = r.updateState(ctx, RayJobInstance, rayv1alpha1.WaitForDashboard, err)
+		return ctrl.Result{}, err
+	}
+
+	rayDashboardClient := utils.RayDashboardClient{}
```

Comment:
minor: I feel the time we initialize the client, it should be ready to use. Seems there's no need to have two lines here?